src/ring_theory/tensor_product.lean

```diff
@@ -1,36 +1,321 @@
 
+open_locale tensor_product
+open tensor_product
+
+-- move this
 namespace algebra
 
-open_locale tensor_product
+variables {R A M N P : Type*}
+variables [comm_semiring R] [semiring A] [algebra R A]
+variables [add_comm_monoid M] [semimodule R M] [semimodule A M] [is_scalar_tower R A M]
+variables [add_comm_monoid N] [semimodule R N]
+
+variables (R M)
+
+/-- The `R`-algebra morphism `A → End (M)` corresponding to the `A`-module structure on `M`. -/
+def lsmul : A →ₐ[R] (M →ₗ[R] M) :=
+{ map_one' := by { ext m, exact one_smul A m },
+  map_mul' := by { intros a b, ext c, exact smul_assoc a b c },
+  map_zero' := by { ext m, exact zero_smul A m },
+  commutes' := by { intro r, ext m, exact algebra_map_smul A r m },
+  .. (show A →ₗ[R] M →ₗ[R] M, from linear_map.mk₂ R (•)
+  (λ x y z, add_smul x y z)
+  (λ c x y, smul_assoc c x y)
+  (λ x y z, smul_add x y z)
+  (λ c x y, smul_algebra_smul_comm c x y)) }
+
+@[simp] lemma lsmul_coe (a : A) : (lsmul R M a : M → M) = (•) a := rfl
+
+instance : is_scalar_tower R A A :=
+{ smul_assoc := λ r a b, by simp only [smul_eq_mul, smul_mul_assoc] }
+
+end algebra
+
+namespace tensor_product
+
+variables {R A M N P : Type*}
+
+section restrict_scalars
+variables [comm_semiring R] [comm_semiring A] [algebra R A]
+variables [add_comm_monoid M] [semimodule A M]
+variables [add_comm_monoid N] [semimodule A N]
+
+variables (R A M N)
+
+@[priority 500]
+instance restrict_scalars.semimodule : semimodule R (M ⊗[A] N) :=
+show semimodule R (restrict_scalars R A (M ⊗ N)), from
+restrict_scalars.semimodule R A (M ⊗ N)
+
+instance restrict_scalars.is_scalar_tower : is_scalar_tower R A (M ⊗[A] N) :=
+show is_scalar_tower R A (restrict_scalars R A (M ⊗ N)), from
+restrict_scalars.is_scalar_tower R A (M ⊗ N)
+
+variables [semimodule R M] [semimodule R N]
+
+/- Check that we didn't introduce a diamond. -/
+example : tensor_product.restrict_scalars.semimodule R R M N =
+          @tensor_product.semimodule R _ M N _ _ _ _ := rfl
+
+end restrict_scalars
+
+/-!
+### The `A`-module structure on `A ⊗[R] M`
+-/
+
+open linear_map
+open algebra (lsmul)
+
+section defn
+variables [comm_semiring R] [semiring A] [algebra R A]
+variables [add_comm_monoid M] [semimodule R M] [semimodule A M] [is_scalar_tower R A M]
+variables [add_comm_monoid N] [semimodule R N]
+
+variables (R A M N)
+
+@[priority 500]
+instance algebra_tensor_module : semimodule A (M ⊗[R] N) :=
+{ smul := λ a, (lsmul R M a).rtensor N,
+  one_smul := λ x, by simp only [alg_hom.map_one, one_def, id.def, id_coe, rtensor_id],
+  mul_smul := λ a b x, by simp only [alg_hom.map_mul, mul_def, rtensor_comp, comp_apply],
+  smul_add := λ a x y, by simp only [map_add],
+  smul_zero := λ a, by simp only [map_zero],
+  add_smul := λ a b x, by simp only [alg_hom.map_add, add_apply, rtensor_add],
+  zero_smul := λ x, by simp only [rtensor_zero, alg_hom.map_zero, zero_apply] }
+
+/- Check that we didn't introduce a diamond. -/
+example : tensor_product.algebra_tensor_module R R M N =
```

Comment:
Same comment here, should this be in the `tests` folder

Reply:
@jcommelin 's reply to an equivalent comment (now hidden) https://github.com/leanprover-community/mathlib/pull/4773#discussion_r526629293 was:
```
I would like to keep it here. It doesn't take up a lot of spacetime, and it is an important check that people reading/using this file might be worried about.
```